include/swift/AST/AutoDiff.h

```diff
@@ -101,145 +101,41 @@ class ParsedAutoDiffParameter {
-
-private:
-  AutoDiffParameterIndices(const llvm::SmallBitVector &parameters)
-      : parameters(parameters) {}
-
-public:
-  /// Allocates and initializes an `AutoDiffParameterIndices` corresponding to
-  /// the given `string` generated by `getString()`. If the string is invalid,
-  /// returns nullptr.
-  static AutoDiffParameterIndices *create(ASTContext &C, StringRef string);
-
-  /// Returns a textual string description of these indices,
-  ///
-  ///   [SU]+
-  ///
-  /// "S" means that the corresponding index is set
-  /// "U" means that the corresponding index is unset
-  std::string getString() const;
-
-  /// Tests whether this set of parameters is empty.
-  bool isEmpty() const { return parameters.none(); }
-
-  /// Pushes the subset's parameter's types to `paramTypes`, in the order in
-  /// which they appear in the function type. For example,
-  ///
-  ///   functionType = (A, B, C) -> R
-  ///   if "A" and "C" are in the set,
-  ///   ==> pushes {A, C} to `paramTypes`.
-  ///
-  ///   functionType = (A, B) -> (C, D) -> R
-  ///   if "A", "C", and "D" are in the set,
-  ///   ==> pushes {A, C, D} to `paramTypes` if `reverseCurryLevels` is false,
-  ///    or pushes {C, D, A} otherwise.
-  ///
-  ///   functionType = (Self) -> (A, B, C) -> R
-  ///   if "Self" and "C" are in the set,
-  ///   ==> pushes {Self, C} to `paramTypes` if `reverseCurryLevels` is false,
-  ///    or pushes {C, Self} otherwise.
-  ///
-  void getSubsetParameterTypes(AnyFunctionType *functionType,
-                               SmallVectorImpl<Type> &paramTypes,
-                               bool reverseCurryLevels = false) const;
-
-  /// Returns a bitvector for the SILFunction parameters corresponding to the
-  /// parameters in this set. In particular, this explodes tuples. For example,
-  ///
-  ///   functionType = (A, B, C) -> R
-  ///   if "A" and "C" are in the set,
-  ///   ==> returns 101
-  ///   (because the lowered SIL type is (A, B, C) -> R)
-  ///
-  ///   functionType = (Self) -> (A, B, C) -> R
-  ///   if "Self" and "C" are in the set,
-  ///   ==> returns 0011
-  ///   (because the lowered SIL type is (A, B, C, Self) -> R)
-  ///
-  ///   functionType = (A, (B, C), D) -> R
-  ///   if "A" and "(B, C)" are in the set,
-  ///   ==> returns 1110
-  ///   (because the lowered SIL type is (A, B, C, D) -> R)
-  ///
-  AutoDiffIndexSubset *getLowered(ASTContext &ctx,
-                                  AnyFunctionType *functionType) const;
-
-  void Profile(llvm::FoldingSetNodeID &ID) const {
-    ID.AddInteger(parameters.size());
-    for (unsigned setBit : parameters.set_bits())
-      ID.AddInteger(setBit);
-  }
-};
+class AnyFunctionType;
+class AutoDiffIndexSubset;
+class Type;
+class SILModule;
+enum class SILLinkage : uint8_t;
 
-/// Builder for `AutoDiffParameterIndices`.
-class AutoDiffParameterIndicesBuilder {
+/// Builder for `AutoDiffIndexSubset`.
+class AutoDiffIndexSubsetBuilder {
```

Comment:
I don't think we need a `AutoDiffIndexSubsetBuilder`. `AutoDiffIndexSubset` is supposed to be a lightweight data structure that can be created using its factory method. Moreover, Sema is the only user of it. Instead, you should just create a `SmallBitVector`, which allows you to set/unset bits in place, and call `AutoDiffIndexSubset::get` on it to create an index subset when you've done modifying in place.

Reply:
Nuked `AutoDiffIndexSubsetBuilder` in https://github.com/apple/swift/pull/27514/commits/d093e95afb3d67e94c3aa1efd697a97141b02c36.